File: reverse/game2/3-readable/tjge/GameMIDlet.java
// 派生文件 · 非权威 · 逻辑逐字节等同 2-decompiled-cfr/tjge/GameMIDlet.java
// 标识符按 reverse/game2/3-readable/SYMBOLS.md 重命名，仅供阅读；任何不一致以 CFR 为准。
/*
 * Decompiled with CFR 0.152.
 *
 * Could not load the following classes:
 *  javax.microedition.lcdui.Display
 *  javax.microedition.lcdui.Displayable
 *  javax.microedition.lcdui.Image
 *  javax.microedition.media.Manager
 *  javax.microedition.media.Player
 *  javax.microedition.media.control.VolumeControl
 *  javax.microedition.midlet.MIDlet
 *  javax.microedition.midlet.MIDletStateChangeException
 *  javax.microedition.rms.RecordEnumeration
 *  javax.microedition.rms.RecordStore
 */
package tjge;

import java.io.IOException;
import java.io.InputStream;
import java.util.Random;
import javax.microedition.lcdui.Display;
import javax.microedition.lcdui.Displayable;
import javax.microedition.lcdui.Image;
import javax.microedition.media.Manager;
import javax.microedition.media.Player;
import javax.microedition.media.control.VolumeControl;
import javax.microedition.midlet.MIDlet;
import javax.microedition.midlet.MIDletStateChangeException;
import javax.microedition.rms.RecordEnumeration;
import javax.microedition.rms.RecordStore;
import tjge.GameCanvas;

public class GameMIDlet
extends MIDlet {
    GameCanvas canvas = new GameCanvas(this);                       // 字段 a：主游戏画布（实例）
    Display display = Display.getDisplay((MIDlet)this);             // 字段 b：LCDUI 显示管理器
    public static Random random;                                   // 字段 c：全局随机数发生器
    static byte[] byteBuf1;                                         // 字段 d：1 字节读缓冲
    static byte[] byteBuf2;                                         // 字段 e：2 字节读缓冲
    static byte[] byteBuf4;                                         // 字段 f：4 字节读缓冲
    public static int soundTimeout;                                // 字段 g：音轨回收倒计时
    public static int soundTrackCount;                             // 字段 h：音轨数量（=2）
    public static Player[] soundPlayers;                           // 字段 i：MIDI 播放器数组
    public static int currentSoundIndex;                           // 字段 j：当前播放音轨索引，-1 表示无
    public static byte[] saveRecord;                               // 字段 k：5 字节存档记录，k[2]=声音开关
    static String tempText1;                                       // 字段 l：临时文本串 1
    static String tempText2;                                       // 字段 m：临时文本串 2
    static String tempText3;                                       // 字段 n：临时文本串 3
    public static final String[] menuTexts;                        // 字段 o：菜单文案表
    public static final String[] interludeTexts;                   // 字段 p：过场/结算文案表
    public static final String[] missionTexts;                     // 字段 q：任务目标文案表
    public static final String[] numeralTexts;                     // 字段 r：关卡序号文案表

    // 构造函数：初始化全局 Random 并以当前时间设种子。
    public GameMIDlet() {
        random = new Random();
        random.setSeed(System.currentTimeMillis());
    }

    // MIDlet 暂停回调：若画布处于关卡态(uiState==10)则切回菜单态并清相关计数。
    protected final void pauseApp() {
        if (this.canvas.uiState == 10) {
            this.canvas.inputAction = 0;
            this.canvas.uiState = 4;
            this.canvas.menuStartItem = 0;
            this.canvas.menuSelection = 0;
        }
    }

    // MIDlet 启动回调：将 GameCanvas 设为当前可显示对象。
    protected final void startApp() throws MIDletStateChangeException {
        this.display.setCurrent((Displayable)this.canvas);
    }

    // MIDlet 销毁回调：持久化存档、清空显示与画布引用并通知销毁。
    protected final void destroyApp(boolean bl) throws MIDletStateChangeException {
        GameMIDlet.accessSaveRecord(1);
        this.display.setCurrent(null);
        this.canvas.running = false;
        this.canvas.mainThread = null;
        this.notifyDestroyed();
    }

    // a(I)I → randomBelow：返回 [0,limit) 区间的随机整数（取随机数绝对值后取模）。
    public static final int randomBelow(int limit) {
        int value = random.nextInt();
        if (value < 0) {
            value = -value;
        }
        return value % limit;
    }

    // a(Ljava/io/InputStream;)B → readByte：从输入流读 1 字节并返回（有符号）。
    public static final byte readByte(InputStream inputStream) throws IOException {
        inputStream.read(byteBuf1);
        return byteBuf1[0];
    }

    // b(Ljava/io/InputStream;)S → readShortLE：从输入流读小端 16 位 short。
    public static final short readShortLE(InputStream inputStream) throws IOException {
        inputStream.read(byteBuf2);
        int value = byteBuf2[1];
        value <<= 8;
        return (short)(value |= byteBuf2[0] & 0xFF);
    }

    // a([BII)I → readIntLE：从字节数组按小端读取 count 字节组成 int（最高字节有符号）。
    public static final int readIntLE(byte[] byArray, int offset, int count) {
        int result = 0;
        int idx = count - 1;
        while (idx >= 0) {
            result <<= 8;
            result = idx == count - 1 ? (result |= byArray[idx + offset]) : (result |= byArray[idx + offset] & 0xFF);
            --idx;
        }
        return result;
    }

    // a()V → tickSoundTimeout：声音超时回收：倒计时 soundTimeout 到期且曲终则反分配当前音轨。
    public static final void tickSoundTimeout() {
        if (currentSoundIndex >= 0 && --soundTimeout < 0 && soundPlayers[currentSoundIndex].getState() != 400) {
            soundPlayers[currentSoundIndex].deallocate();
            currentSoundIndex = -1;
        }
    }

    // b()V → loadSounds：加载 sound.bin 中的各条 MIDI 创建 Player 数组并设音量。
    public static final void loadSounds() {
        int index = 0;
        while (index < soundTrackCount) {
            try {
                InputStream inputStream = GameMIDlet.openEntryStream("/res/sound.bin", index);
                GameMIDlet.soundPlayers[index] = Manager.createPlayer((InputStream)inputStream, (String)"audio/midi");
                soundPlayers[index].realize();
                VolumeControl volumeControl = (VolumeControl)soundPlayers[index].getControl("VolumeControl");
                if (volumeControl != null) {
                    volumeControl.setLevel(90);
                }
                inputStream.close();
            }
            catch (Exception exception) {}
            ++index;
        }
    }

    // a(II)V → playSound：按声音开关(saveRecord[2])单次播放第 trackIndex 条音轨并设回收倒计时。
    public static final void playSound(int trackIndex, int unused) {
        if (saveRecord[2] == 0) {
            return;
        }
        try {
            GameMIDlet.stopSound();
            soundPlayers[trackIndex].setLoopCount(1);
            soundPlayers[trackIndex].start();
            soundTimeout = 2;
            currentSoundIndex = trackIndex;
            return;
        }
        catch (Exception exception) {
            return;
        }
    }

    // c()V → stopSound：停止当前正在播放的音轨。
    public static final void stopSound() {
        if (currentSoundIndex < 0) {
            return;
        }
        try {
            if (soundPlayers[currentSoundIndex].getState() == 400) {
                soundPlayers[currentSoundIndex].stop();
            }
            return;
        }
        catch (Exception exception) {
            return;
        }
    }

    // b(I)V → accessSaveRecord：RMS 存档读写：mode=0 读取/1 写入/2 清零前两字节(REDDEVIL2,5字节)。
    public static final void accessSaveRecord(int mode) {
        try {
            if (mode == 2) {
                GameMIDlet.saveRecord[0] = 0;
                GameMIDlet.saveRecord[1] = 0;
                return;
            }
            RecordStore recordStore = RecordStore.openRecordStore((String)"REDDEVIL2", (boolean)true);
            RecordEnumeration recordEnumeration = recordStore.enumerateRecords(null, null, false);
            if (recordEnumeration.hasNextElement()) {
                int recordId = recordEnumeration.nextRecordId();
                if (mode == 0) {
                    saveRecord = recordStore.getRecord(recordId);
                } else if (mode == 1) {
                    recordStore.setRecord(recordId, saveRecord, 0, 5);
                }
            } else if (mode == 0) {
                GameMIDlet.saveRecord[0] = 0;
                GameMIDlet.saveRecord[1] = 0;
            } else if (mode == 1) {
                recordStore.addRecord(saveRecord, 0, 5);
            }
            recordEnumeration.destroy();
            recordStore.closeRecordStore();
            return;
        }
        catch (Exception exception) {
            return;
        }
    }

    // a(ILjava/lang/String;)V → loadTextEntry：从归档第 entryIndex 条读 UTF-16 文本，存入临时串 tempText1(<7) 或 tempText3(>=7)。
    public static final void loadTextEntry(int entryIndex, String archive) {
        byte[] bytes = GameMIDlet.readEntryBytes(archive, entryIndex);
        if (bytes != null) {
            char[] chars = new char[bytes.length / 2];
            int i = 0;
            while (i < bytes.length / 2) {
                chars[i] = (char)GameMIDlet.readIntLE(bytes, i * 2, 2);
                ++i;
            }
            if (entryIndex < 7) {
                tempText1 = new String(chars);
            } else {
                GameMIDlet.tempText3 = new String(chars);
            }
            System.gc();
        }
    }

    // a(Ljava/lang/String;I)Ljavax/microedition/lcdui/Image; → loadImage：从归档第 entryIndex 条取 PNG 字节并创建 Image。
    public static final Image loadImage(String archive, int entryIndex) {
        byte[] bytes = GameMIDlet.readEntryBytes(archive, entryIndex);
        Image image = Image.createImage((byte[])bytes, (int)0, (int)bytes.length);
        System.gc();
        return image;
    }

    // b(Ljava/lang/String;I)Ljava/io/InputStream; → openEntryStream：返回定位到归档第 entryIndex 条目起始偏移的输入流。
    static final InputStream openEntryStream(String archive, int entryIndex) {
        int entryOffset = 0;
        try {
            InputStream inputStream = archive.getClass().getResourceAsStream(archive);
            inputStream.read(byteBuf4);
            int entryCount = GameMIDlet.readIntLE(byteBuf4, 0, 4);
            int i = 0;
            while (i < entryCount) {
                inputStream.read(byteBuf4);
                if (i == entryIndex) {
                    entryOffset = GameMIDlet.readIntLE(byteBuf4, 0, 4);
                }
                ++i;
            }
            inputStream.skip(entryOffset);
            return inputStream;
        }
        catch (Exception exception) {
            return null;
        }
    }

    // c(Ljava/lang/String;I)[B → readEntryBytes：读取归档第 entryIndex 条目的完整字节切片。
    static final byte[] readEntryBytes(String archive, int entryIndex) {
        int entryOffset = 0;
        int entryLength = 0;
        try {
            byte[] bytes = null;
            InputStream inputStream = archive.getClass().getResourceAsStream(archive);
            inputStream.read(byteBuf4);
            int entryCount = GameMIDlet.readIntLE(byteBuf4, 0, 4);
            int i = 0;
            while (i < entryCount) {
                inputStream.read(byteBuf4);
                if (i == entryIndex) {
                    entryOffset = GameMIDlet.readIntLE(byteBuf4, 0, 4);
                } else if (i == entryIndex + 1) {
                    entryLength = GameMIDlet.readIntLE(byteBuf4, 0, 4) - entryOffset;
                }
                ++i;
            }
            inputStream.skip(entryOffset);
            bytes = new byte[entryLength];
            inputStream.read(bytes);
            inputStream.close();
            return bytes;
        }
        catch (Exception exception) {
            return null;
        }
    }

    // <clinit>()V → staticInit：静态初始化：建读缓冲、音轨数/数组、默认存档及全部文案表。
    static {
        byteBuf1 = new byte[1];
        byteBuf2 = new byte[2];
        byteBuf4 = new byte[4];
        soundTrackCount = 2;
        soundPlayers = new Player[soundTrackCount];
        currentSoundIndex = -1;
        saveRecord = new byte[]{0, 0, 1, 0, 6};
        menuTexts = new String[]{"返回游戏", "新游戏", "继续", "声音 开", "帮助", "关于", "退出", "声音 关"};
        interludeTexts = new String[]{"继续", "跳过", "任务", "返回", "完成", "击毙敌人:", "所用时间:", "继续次数:", "失败", "菜单", "游戏结束", "载入中", "."};
        missionTexts = new String[]{"找到船舱入口。", "找到罗斯上校。", "控制台输入锁死密码", "摧毁动力装置", "摧毁备用动力装置", "摧毁直升飞机", "摧毁导弹艇"};
        numeralTexts = new String[]{"一", "二", "三", "四", "五", "六", "七", "八", "九", "十"};
    }
}

File: packages/game2/src/GameMIDlet.ts
/**
 * 游戏2《红魔特种兵2-深海战舰》MIDlet 入口。
 * 逐行移植自 reverse/game2/2-decompiled-cfr/tjge/GameMIDlet.java（CFR 权威源）。
 * 移植规约见 docs/05-移植规约.md；方法名映射见 reverse/game2/porting-naming/porting-contract.json。
 *
 * 与游戏1 差异：状态字段为 a.b（非 a.p）；资源 int 经字节数组读取器 a_AYII（小端）；
 *   音频用 javax.microedition.media（MIDI Player）—— 暂以空实现占位（音频后续接入，不影响画面/逻辑一致）；
 *   存档 RMS "REDDEVIL2"(5字节) → localStorage。
 *
 * 别名：c=全局Random, d/e/f=读缓冲, g=音超时, h=音轨数, i=MIDI播放器[], j=当前音轨,
 *   k=5字节存档, l/m/n=临时字符串, o/p/q/r=菜单/过场/任务/数字 文案表。
 *
 * 角色：整个游戏2 的程序入口与"基础设施层"。三大职责——
 *   ① MIDlet 生命周期（startApp/pauseApp/destroyApp）并持有唯一的主 Canvas（{@link GameCanvas}）与 Display；
 *   ② .bin 归档的字节 IO 原语（小端读取、定位流、整条切片、PNG/字符串/MIDI 加载），是全游戏唯一的资源读取入口；
 *   ③ 音频（MIDI）与存档（RMS→localStorage）的全局静态服务。
 * 协作者：构造时创建 {@link GameCanvas}（主循环/渲染/状态机均在其中）；c 是**全游戏唯一随机源**，
 *   被各处的 {@link GameMIDlet.randomBelow} 复用。
 *
 * 存档 k=saveRecord[5] 各字节语义（据 docs「类清单与职责」）：
 *   [0]=当前关、[1]=继续次数、[2]=声音开关(1=开)、[3]=最高已通关、[4]=存档时玩家生命。
 *
 * CFR 基准：reverse/game2/2-decompiled-cfr/tjge/GameMIDlet.java；语义参见
 *   docs/game2-深海战舰/{类清单与职责.md,资源映射.md}。
 */
import {
  MIDlet,
  Display,
  Random,
  InputStream,
  getResourceAsStream,
  getCachedImage,
  Image,
  MidiSynth,
  getAudioContext,
} from "@red-devil/j2me-shim";
import { GameCanvas } from "./GameCanvas.ts";
import { UiState } from "./constants.ts";

export class GameMIDlet extends MIDlet {
  canvas: GameCanvas;
  display: Display;
  static random: Random;
  static byteBuf1: Int8Array = new Int8Array(1);
  static byteBuf2: Int8Array = new Int8Array(2);
  static byteBuf4: Int8Array = new Int8Array(4);
  static soundTimeout = 0;
  static soundTrackCount = 2;
  static soundPlayers: (MidiSynth | null)[] = new Array(GameMIDlet.soundTrackCount).fill(null); // MIDI 合成器[]
  static currentSoundIndex = -1;
  static saveRecord: Int8Array = Int8Array.from([0, 0, 1, 0, 6]); // 5 字节存档
  static tempText1 = "";
  static tempText2 = "";
  static tempText3 = "";
  // 菜单文案：返回游戏/新游戏/继续/声音 开/帮助/关于/退出/声音 关
  static menuTexts: string[] = ["返回游戏", "新游戏", "继续", "声音 开", "帮助", "关于", "退出", "声音 关"];
  // 过场/结算文案
  static interludeTexts: string[] = [
    "继续", "跳过", "任务", "返回", "完成", "击毙敌人:", "所用时间:", "继续次数:", "失败", "菜单", "游戏结束", "载入中", ".",
  ];
  // 任务目标文案
  static missionTexts: string[] = [
    "找到船舱入口。", "找到罗斯上校。", "控制台输入锁死密码", "摧毁动力装置", "摧毁备用动力装置", "摧毁直升飞机", "摧毁导弹艇",
  ];
  // 关卡序号（一~十）
  static numeralTexts: string[] = ["一", "二", "三", "四", "五", "六", "七", "八", "九", "十"];

  /**
   * 构造 MIDlet：创建主 Canvas、取 Display、初始化全局随机源。
   * 严格复刻 Java 求值顺序（字段初始化器先于构造体）：先 `new GameCanvas(this)`、
   * 再 `Display.getDisplay(this)`，最后才是构造体里的 `c = new Random(); c.setSeed(currentTimeMillis())`。
   * c 随后成为全游戏唯一随机源。CFR：GameMIDlet.java 字段 + 构造 54-57。
   */
  constructor() {
    super();
    // Java 字段初始化器先于构造体执行：`a a = new a(this)`、`b = Display.getDisplay(this)`
    // 在构造体之前求值，最后才是构造体里的 c。严格复刻此求值顺序。
    this.canvas = new GameCanvas(this);
    this.display = Display.getDisplay(this);
    GameMIDlet.random = new Random();
    GameMIDlet.random.setSeed(Date.now());
  }

  /**
   * MIDlet 暂停回调：若正处于游戏中（InGame），清输入动作并复位菜单状态、强制切回主菜单。
   * 复刻原 pauseApp（来电/切后台时让游戏挂起到菜单）。CFR：GameMIDlet.java pauseApp。
   */
  pauseApp(): void {
    if (this.canvas.uiState === UiState.InGame) {
      this.canvas.inputAction = 0;
      this.canvas.uiState = UiState.MainMenu;
      this.canvas.menuStartItem = 0;
      this.canvas.menuSelection = 0;
    }
  }

  /**
   * MIDlet 启动回调：把主 Canvas 设为当前显示，进入游戏。CFR：GameMIDlet.java startApp。
   */
  startApp(): void {
    this.display.setCurrent(this.canvas);
  }

  /**
   * MIDlet 销毁回调：写存档（{@link GameMIDlet.accessSaveRecord} 传 1=写）、解绑显示、
   * 停主循环线程，最后 notifyDestroyed()。CFR：GameMIDlet.java destroyApp。
   */
  destroyApp(_unconditional: boolean): void {
    GameMIDlet.accessSaveRecord(1);
    this.display.setCurrent(null);
    this.canvas.running = false;
    this.canvas.mainThread = null as unknown as GameCanvas["mainThread"];
    this.notifyDestroyed();
  }

  /** GameMIDlet.a(int)：随机数绝对值取模，返回 [0,bound) 内的值。 */
  static randomBelow(bound: number): number {
    let value = GameMIDlet.random.nextInt();
    if (value < 0) value = -value;
    return value % bound;
  }

  /** GameMIDlet.a(InputStream)：读 1 字节（有符号）。 */
  static readByte(inputStream: InputStream): number {
    inputStream.read(GameMIDlet.byteBuf1);
    return GameMIDlet.byteBuf1[0];
  }

  /** GameMIDlet.b(InputStream)：读小端 16 位（高字节有符号，`(x<<16)>>16` 复现 Java short 截断）。 */
  static readShortLE(inputStream: InputStream): number {
    inputStream.read(GameMIDlet.byteBuf2);
    let value = GameMIDlet.byteBuf2[1];
    value <<= 8;
    value |= GameMIDlet.byteBuf2[0] & 0xff;
    return (value << 16) >> 16;
  }

  /** GameMIDlet.a(byte[],int,int)：从 bytes 偏移 offset 按小端读 byteCount 字节为 int（最高字节有符号，低字节 `&0xff` 去符号）。 */
  static readIntLE(bytes: Int8Array, offset: number, byteCount: number): number {
    let result = 0;
    let i = byteCount - 1;
    while (i >= 0) {
      result <<= 8;
      if (i === byteCount - 1) result |= bytes[i + offset];
      else result |= bytes[i + offset] & 0xff;
      --i;
    }
    return result | 0;
  }

  // ---- 音频（MIDI）：自带轻量 WebAudio 合成器播放 sound.bin 的标准 MIDI（设备相关，最佳努力）----
  /** 原 a()：单次播放结束后的超时回收（g 倒计时，曲终且 g 到期则释放）。 */
  static tickSoundTimeout(): void {
    if (GameMIDlet.currentSoundIndex >= 0 && --GameMIDlet.soundTimeout < 0) {
      const p = GameMIDlet.soundPlayers[GameMIDlet.currentSoundIndex];
      if (p && !p.isPlaying()) {
        p.stop();
        GameMIDlet.currentSoundIndex = -1;
      }
    }
  }
  /** 原 b()：加载 sound.bin 的 MIDI 到合成器数组。 */
  static loadSounds(): void {
    const ctx = getAudioContext();
    if (!ctx) return;
    let trackIndex = 0;
    while (trackIndex < GameMIDlet.soundTrackCount) {
      try {
        const bytes = GameMIDlet.readEntryBytes("/res/sound.bin", trackIndex);
        if (bytes != null) {
          GameMIDlet.soundPlayers[trackIndex] = new MidiSynth(new Uint8Array(bytes.buffer, bytes.byteOffset, bytes.byteLength), ctx);
        }
      } catch (exception) {
        /* ignore */
      }
      ++trackIndex;
    }
  }
  /** 原 a(int,int)：播放音轨 trackIndex（k[2]=声音开关；setLoopCount(1)=单次）。_n2 为保留匹配原签名的死参。 */
  static playSound(trackIndex: number, _n2: number): void {
    if (GameMIDlet.saveRecord[2] === 0) return; // 声音关
    try {
      GameMIDlet.stopSound();
      const p = GameMIDlet.soundPlayers[trackIndex];
      if (p) {
        p.play(false); // 原 setLoopCount(1)：单次播放
        GameMIDlet.soundTimeout = 2;
        GameMIDlet.currentSoundIndex = trackIndex;
      }
    } catch (exception) {
      /* ignore */
    }
  }
  /** 原 c()：停止当前音轨。 */
  static stopSound(): void {
    if (GameMIDlet.currentSoundIndex < 0) return;
    try {
      GameMIDlet.soundPlayers[GameMIDlet.currentSoundIndex]?.stop();
    } catch (exception) {
      /* ignore */
    }
  }

  /**
   * GameMIDlet.b(int)：RMS 存档 "REDDEVIL2"(5 字节 [关/继续数/声音/进度/储备弹]) ↔ localStorage。
   * @param mode 0=读入 saveRecord；1=写出 saveRecord；2=清头两字节（关卡/继续数归零，不落盘）
   */
  static accessSaveRecord(mode: number): void {
    try {
      if (mode === 2) {
        GameMIDlet.saveRecord[0] = 0;
        GameMIDlet.saveRecord[1] = 0;
        return;
      }
      const raw = typeof localStorage !== "undefined" ? localStorage.getItem("REDDEVIL2") : null;
      if (raw) {
        if (mode === 0) {
          const arr = raw.split(",").map((x) => Number(x));
          GameMIDlet.saveRecord = Int8Array.from(arr);
        } else if (mode === 1) {
          localStorage.setItem("REDDEVIL2", Array.from(GameMIDlet.saveRecord.slice(0, 5)).join(","));
        }
      } else {
        if (mode === 0) {
          GameMIDlet.saveRecord[0] = 0;
          GameMIDlet.saveRecord[1] = 0;
        } else if (mode === 1) {
          if (typeof localStorage !== "undefined") {
            localStorage.setItem("REDDEVIL2", Array.from(GameMIDlet.saveRecord.slice(0, 5)).join(","));
          }
        }
      }
    } catch (exception) {
      /* ignore */
    }
  }

  /** GameMIDlet.a(int,String)：从 archivePath 第 entryIndex 条读 UTF-16 文本，存入 tempText1(entryIndex<7) 或 tempText3。 */
  static loadTextEntry(entryIndex: number, archivePath: string): void {
    const bytes = GameMIDlet.readEntryBytes(archivePath, entryIndex);
    if (bytes != null) {
      const chars = new Array<number>(bytes.length / 2);
      let i = 0;
      while (i < ((bytes.length / 2) | 0)) {
        chars[i] = GameMIDlet.readIntLE(bytes, i * 2, 2);
        ++i;
      }
      const text = String.fromCharCode(...chars);
      if (entryIndex < 7) GameMIDlet.tempText1 = text;
      else GameMIDlet.tempText3 = text;
      // System.gc();
    }
  }

  /** GameMIDlet.a(String,int)：从 archivePath 第 entryIndex 条 PNG 取图（预解码缓存）。 */
  static loadImage(archivePath: string, entryIndex: number): Image {
    // 原版：c(archivePath,entryIndex) 取 PNG 字节 → Image.createImage；这里取预解码缓存（见 docs/05 §5）
    return getCachedImage<Image>(archivePath, entryIndex);
  }

  /** GameMIDlet.b(String,int)：返回定位到 archivePath 第 entryIndex 条起始的输入流。 */
  static openEntryStream(archivePath: string, entryIndex: number): InputStream | null {
    let offset = 0;
    try {
      const inputStream = getResourceAsStream(archivePath)!;
      inputStream.read(GameMIDlet.byteBuf4);
      const entryCount = GameMIDlet.readIntLE(GameMIDlet.byteBuf4, 0, 4);
      let i = 0;
      while (i < entryCount) {
        inputStream.read(GameMIDlet.byteBuf4);
        if (i === entryIndex) offset = GameMIDlet.readIntLE(GameMIDlet.byteBuf4, 0, 4);
        ++i;
      }
      inputStream.skip(offset);
      return inputStream;
    } catch (exception) {
      return null;
    }
  }

  /** GameMIDlet.c(String,int)：取 archivePath 第 entryIndex 条目的字节切片（读第 entryIndex 与 entryIndex+1 条 offset 之差为长度）。 */
  static readEntryBytes(archivePath: string, entryIndex: number): Int8Array | null {
    let startOffset = 0;
    let byteCount = 0;
    try {
      const inputStream = getResourceAsStream(archivePath)!;
      inputStream.read(GameMIDlet.byteBuf4);
      const entryCount = GameMIDlet.readIntLE(GameMIDlet.byteBuf4, 0, 4);
      let i = 0;
      while (i < entryCount) {
        inputStream.read(GameMIDlet.byteBuf4);
        if (i === entryIndex) startOffset = GameMIDlet.readIntLE(GameMIDlet.byteBuf4, 0, 4);
        else if (i === entryIndex + 1) byteCount = GameMIDlet.readIntLE(GameMIDlet.byteBuf4, 0, 4) - startOffset;
        ++i;
      }
      inputStream.skip(startOffset);
      const result = new Int8Array(byteCount);
      inputStream.read(result);
      inputStream.close();
      return result;
    } catch (exception) {
      return null;
    }
  }
}
